src/core/server/http/http_server.ts

```diff
@@ -323,17 +331,32 @@ export class HttpServer {
     this.server.events.on('response', this.handleServerResponseEvent);
   }
 
-  private setupRequestStateAssignment(config: HttpConfig) {
+  private setupRequestStateAssignment(
+    config: HttpConfig,
+    executionContext?: InternalExecutionContextSetup
+  ) {
     this.server!.ext('onRequest', (request, responseToolkit) => {
+      const requestId = getRequestId(request, config.requestId);
+
+      const parentContext = executionContext?.getParentContextFrom(request.headers);
+      executionContext?.set({ ...parentContext, requestId });
+
       request.app = {
         ...(request.app ?? {}),
-        requestId: getRequestId(request, config.requestId),
+        requestId,
         requestUuid: uuid.v4(),
       } as KibanaRequestState;
       return responseToolkit.continue;
     });
   }
 
+  private setupContextExecutionCleanup(executionContext?: InternalExecutionContextSetup) {
+    if (!executionContext) return;
+    this.server!.events.on('response', function () {
```

Comment:
From the docs: 

https://hapi.dev/api/?v=20.1.4#-serverextevents
> server.ext
Registers an extension function in one of the request lifecycle extension points 

https://hapi.dev/api/?v=20.1.4#-request-event
> server.events.on('request'
> The 'request' event type emits internal request events generated by the framework as well as application events logged with request.log()

This, `server.events.on('request'` has a different semantic of `hapi internal request events` and it can be emitted several times for an incoming request.